interface/MO_Interface.h

```diff
@@ -0,0 +1,173 @@
+   * implementing this interface must ensure that it will call the DMClient::XXX() method 
+   * exactly once for each time this callback is invoked by the client library and the
+   * callback returned true.
+   *
+   * For each invocation of the execute callback by the client library the implementation
+   * can call DMClient::XXX() either from inside its implementation of the callback
+   * for synchronous reporting, or at *any* time after returning from the callback 
+   * (see section 5.4.1 of the OMA DM 2.0 protocol specification) 
+   * 
+   * TODO: check if additional info (parameters) are needed in order
+   * to correctly link the response (synchronous or asynchronous) with the command.
+   * 
+   * This interface is not considered stable yet
+   */
+  virtual bool execute(const std::string node_path) = 0;
+
+
+  /**
+   *  @}
+   *  @{
+   *  (2) callbacks called by the protocol client library as part of Session or MO management
+   */
+
+  /**
+   * @brief callback for the protocol client library to check MO compatibility
+   *
+   * The protocol client library will call this before adding an instance of an MO implementation
+   * to its MO tree in order to check if the given implementation of the MO class is compatible with
+   * the DDF file registered with the client library. 
+   *
+   * @param[in] ddfname - DDFName as given in the respective name in a DDF file. Example: "urn:oma:mo:oma-dm-dmacc:1.2"
+   * @return - MO implementation classes should only return true if they are compatible with the DDF file corresponding to the given DDFName
+   * 
+   */
+  virtual bool check_ddf_name_compatibility(std::string ddfname) = 0;
+
+  /**
+   * @brief callback on adding the MO instance to the client library's MO tree
+   *
+   * The protocol client library will call this after successfully adding the instance of the MO
+   * into its MO tree.
+   */
+  virtual void init_mo() = 0; // gets called by the client after adding an instance to the tree.
+
+  /**
+   * @brief callback on removing the MO instance from the client library's MO tree
+   *
+   *  Use currently unclear. May be removed later if not needed
+   */
+  virtual void close_mo() = 0; // unclear
+
+  /**
+   * @brief callback to read the miid of the MO instance
+   *
+   * The protocol client library will call this to retrieve the miid of a given MO instance.
+   * The MO classes implementing this interface are responsible for ensuring that miid's are 
+   * unique between all instances of this MO type.
+   */
+  virtual std::string miid() const = 0;
+
+  /**
+   * @provide json object for MO content serialization
+
+   * This shall return a JSON object suitable for use in serialization according to the format
+   * described in OMA-DM Protocol Spec V2.0 section 7.2.1.4 - management object serialization.
+   * The output of this method shall correspond to the "MOData" object in that specification.
+   */
+  virtual nlohmann::json serialize_json() const = 0;
+
+  /**
```

Comment:
Do you plan to support "SUB"/"UNSUB" command?

Reply:
Not for the first iteration. SUB/UNSUB are optional for the client to support ("SHOULD" in the table in chapter 5.4 of protocol spec) I can do without them for now.
Long term, since I made a decision to keep the data for the MO tree in the MO implementation (and thus in the local app), I plan for MOs who want to support SUB to call into an interface of the sdk for any node value change. The sdk will only manage the ongoing subscriptions, and decide if any node change reported by MO matches a subscription.
Can you think of alternative approaches, always considering the data store is in the MO/local app, not in the SDK